src/cmap/test/mod.rs

```diff
@@ -191,40 +194,37 @@ impl Operation {
                     state.threads.write().await.insert(target, task.unwrap());
                 }
                 Operation::Wait { ms } => RUNTIME.delay_for(Duration::from_millis(ms)).await,
-                Operation::WaitForThread { target } => state
-                    .threads
-                    .write()
-                    .await
-                    .remove(&target)
-                    .unwrap()
-                    .await
-                    .expect("polling the future should not fail")?,
+                Operation::WaitForThread { target } => {
+                    state.threads.write().await.remove(&target).unwrap().await?
+                }
                 Operation::WaitForEvent { event, count } => {
                     while state.count_events(&event) < count {
                         RUNTIME.delay_for(Duration::from_millis(100)).await;
                     }
                 }
                 Operation::CheckOut { label } => {
                     if let Some(pool) = state.pool.read().await.deref() {
-                        let mut conn = pool.check_out().await?;
+                        let conn = pool.check_out().await?;
 
                         if let Some(label) = label {
                             state.connections.write().await.insert(label, conn);
                         } else {
-                            conn.pool.take();
+                            state.unlabeled_connections.lock().await.push(conn);
                         }
                     }
                 }
                 Operation::CheckIn { connection } => {
                     let conn = state.connections.write().await.remove(&connection).unwrap();
+                    drop(conn);
 
-                    if let Some(pool) = state.pool.read().await.deref() {
-                        pool.check_in(conn).await;
-                    }
+                    // give some time for drop to happen.
+                    RUNTIME.delay_for(Duration::from_millis(500)).await;
```

Comment:
Is there a way we can use channels here to avoid having to sleep and hope that the background task runs fast enough?

Reply:
Definitely, that is what is being done in #258. I didn't want to duplicate the changes from that in this PR though, so I just added heavy wait times. Once that PR goes in I'll rebase this one to use the channel approach.

Reply:
Sounds good!